File: slides/cpu_friendly_code/Demo/branch_prediction.js
var Timer = require('./common/timer.js');

function rand(d) {
    return Math.random() * d;
}

var timer = new Timer;

var TESTS = 10000;
var N = 1000;

var tests = new Array(TESTS);

for (i=0; i<TESTS; ++i) {
    var test = new Array(N);

    for (j=0; j<N; ++j) {
        test[j] = rand(10);
    }

    tests[i] = test;
}

var unsorted_sanity_check = 0;

timer.profile("unsorted", function () {
    for (i=0; i<TESTS; ++i) {
        var test = tests[i];

        for (j=0; j<N; ++j) {
            if (test[j] < 5) {
                unsorted_sanity_check += test[j];
            }
        }
    }
});

for (i=0; i<TESTS; ++i) {
    tests[i].sort();
}

var sorted_sanity_check = 0;

timer.profile("sorted", function () {
    for (i=0; i<TESTS; ++i) {
        var test = tests[i];

        for (j=0; j<N; ++j) {
            if (test[j] < 5) {
                sorted_sanity_check += test[j];
            }
        }
    }
});

console.log(unsorted_sanity_check);
console.log(sorted_sanity_check);

File: slides/cpu_friendly_code/Demo/branch_prediction.rb
require 'hitimes'

class Timer
  def self.profile(label)
    time = Hitimes::Interval.measure do
      yield
    end

    puts "#{label}: #{time * 1000.0} ms"
  end
end

TESTS = 1000
N = 1000

tests = Array.new(TESTS)
sorted_tests = Array.new(TESTS)

TESTS.times do |i|
  test = Array.new(N)

  N.times do |j|
    test[j] = (Random.rand * 10).floor
  end

  tests[i] = test
  sorted_tests[i] = test.sort
end

usum = 0

Timer.profile("unsorted") do
  tests.each do |test|
    test.each do |t|
      usum += t if t < 5
    end
  end
end

ssum = 0

Timer.profile("sorted") do
  sorted_tests.each do |test|
    test.each do |t|
      ssum += t if t < 5
    end
  end
end

puts usum
puts ssum